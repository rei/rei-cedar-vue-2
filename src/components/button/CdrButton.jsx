import clsx from 'clsx';
import modifier from '../../mixins/modifier';
import fullWidth from '../../mixins/fullWidth';
import size from '../../mixins/size';
import space from '../../mixins/space';
import style from './styles/CdrButton.scss';

export default {
  name: 'CdrButton',
  mixins: [modifier, size, space, fullWidth],
  props: {
    /**
     * Controls render as button or anchor. {button, a}
     */
    tag: {
      type: String,
      default: 'button',
      validator: value => (['button', 'a'].indexOf(value) >= 0) || false,
    },
    /**
     * Sets the button type. {button, submit, reset}
     */
    type: {
      type: String,
      default: 'button',
      validator: value => (['button', 'submit', 'reset'].indexOf(value) >= 0) || false,
    },
    /**
     * Renders an icon-only button. Default slot is disabled. Overrides size and responsiveSize props.
     */
    iconOnly: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-button';
    },
    defaultClass() {
      return this.modifyClassName(this.baseClass, 'primary');
    },
    buttonSizeClass() {
      return !this.iconOnly ? this.sizeClass : null;
    },
    iconClass() {
      const classes = [];

      if (this.$slots.icon && this.$slots.default) {
        /* only add class for buttons with text + icon */
        classes.push(this.modifyClassName(this.baseClass, 'has-icon'));
      }

      if (this.iconOnly) {
        classes.push(this.modifyClassName(this.baseClass, 'icon-only'));
      }
      return classes.join(' ');
    },
  },
  render() {
    const Component = this.tag;
    return (<Component
      class={clsx(this.defaultClass,
        this.style[this.baseClass],
        this.modifierClass,
        this.buttonSizeClass,
        this.fullWidthClass,
        this.iconClass,
        this.space)}
      type={this.tag === 'button' ? this.type : null}
      {...{ on: this.$listeners }}
    >
      {this.$slots.icon}
      {this.$slots.default}
    </Component>);
  },
};
