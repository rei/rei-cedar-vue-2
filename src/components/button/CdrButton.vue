<template>
  <component
    :is="tag"
    :class="[
      modifierClass,
      sizeClass,
      fullWidthClass,
      responsiveClass,
      iconClass,
    ]"
    :type="tag === 'button' ? type : null"
    @click="onClick">
    <!-- @slot for icon -->
    <slot name="icon" />
    <!-- @slot innerHTML on the inside of the button component -->
    <slot />
  </component>
</template>

<script>
import modifier from 'mixinsdir/modifier';

/**
 * Cedar 2 component for button
 *
 * CdrButton will render either a button, or an anchor that looks like a button.
 * As such, the decision to use CdrButton vs CdrAnchor should be made based on what
 * you need the rendered element to look like.
 *
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrButton',
  mixins: [modifier],
  props: {
    /**
     * Controls render as button or anchor. {button, a}
     */
    tag: {
      type: String,
      default: 'button',
      validator: value => (['button', 'a'].indexOf(value) >= 0) || false,
    },
    /**
     * Sets the button type. {button, submit, reset}
     */
    type: {
      type: String,
      default: 'button',
      validator: value => (['button', 'submit', 'reset'].indexOf(value) >= 0) || false,
    },
    /**
     * Adds custom click actions.
     */
    onClick: {
      type: Function,
      default: () => () => null,
    },
    /**
     * Sets width to be 100%.
    */
    fullWidth: {
      type: Boolean,
      default: false,
      validator: value => typeof value === 'boolean',
    },
    /**
     * Sets a static size for the button, which scales padding and text size. {small, medium, large}
     */
    size: {
      type: String,
      default: 'medium',
      validator: value => (['small', 'medium', 'large'].indexOf(value) >= 0) || false,
    },
    /**
     * Render a specific button size at a specific breakpoint. Takes precedence over size and fullWidth.
     * Format is size@breakpoint (ex: large@sm).
     */
    responsiveSize: {
      type: Array,
    },
    /**
     * Renders an icon-only button. Default slot is disabled. Overrides size and responsiveSize props.
     */
    iconOnly: {
      type: Boolean,
      default: false,
    },
    /**
     * Renders an icon-only button with a light fill color for use on dark backgrounds.
     * iconOnly must be true.
     */
    onDark: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    baseClass() {
      return 'cdr-button';
    },
    sizeClass() {
      return !this.iconOnly ? this.modifyClassName(this.baseClass, this.size) : null;
    },
    responsiveClass() {
      const responsiveClass = [];

      if (this.responsiveSize && !this.iconOnly) {
        this.responsiveSize.forEach((val) => {
          responsiveClass.push(this.modifyClassName(this.baseClass, val));
        });
      }

      return responsiveClass.join(' ');
    },
    iconClass() {
      const classes = [];

      if (this.$slots.icon && this.$slots.default) {
        /* only add class for buttons with text + icon */
        classes.push(this.modifyClassName(this.baseClass, 'has-icon'));
      }

      if (this.iconOnly) {
        classes.push(this.modifyClassName(this.baseClass, 'icon-only'));

        if (this.onDark) {
          classes.push(this.modifyClassName(this.baseClass, 'on-dark'));
        }
      }
      return classes.join(' ');
    },
    fullWidthClass() {
      return this.fullWidth && !this.iconOnly ?
        this.modifyClassName(this.baseClass, 'full-width') : null;
    },
  },
};
</script>

<style module>
  @import 'cssdir/settings/_index.pcss';
  @import './styles/CdrButton.pcss';
</style>
