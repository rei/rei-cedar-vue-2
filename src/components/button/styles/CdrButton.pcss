/* ==========================================================================
  # CdrButton

  TOC:

    :Base Styles
      :States
    :Modifiers
      :Style variants
        :Secondary
    :Size variants
      :large
      :medium
      :small
    :full-width
    :icon-only
    :on-dark
    :Breakpoint variants
      :xs
      :sm
      :md
      :lg
========================================================================== */

.cdr-button {
  @include redwood-display-30;

  background-color: $button-background-color;
  box-shadow: $button-box-shadow;
  border: none;
  border-radius: $button-border-radius;
  color: $button-color;
  cursor: pointer;
  display: inline-flex;
  fill: $button-color;
  outline: none;
  padding: $inset-1-x-squish;
  text-align: left;
  text-decoration: none;
  text-transform: none;
  overflow: visible;

  & > :global(&__icon) {
    align-self: center;
    fill: inherit;
    margin-right: $space-quarter-x;
  }

  /* States
    ---------- */

  &:hover,
  &:active,
  &:focus {
    background-color: $button-hover-background-color;
    color: $button-hover-color;
    box-shadow: $button-hover-box-shadow;
    text-decoration: none;
    outline: none;
    outline-offset: 0;
    fill: $button-hover-color;
  }

  &:focus {
    outline: $default-outline;
  }

  &[disabled] {
    background-color: $button-disabled-background-color;
    box-shadow: $button-disabled-box-shadow;
    color: $button-disabled-color;
    cursor: not-allowed;
  }

  /* Modifier style variants
    ========================================================================== */

  /* Secondary
     ========== */

  &--secondary {
    background-color: $button-secondary-background-color;
    box-shadow: $button-secondary-box-shadow;
    color: $button-secondary-color;
    fill: $button-secondary-color;

    /* States
    ---------- */

    &:hover,
    &:active,
    &:focus {
      background-color: $button-secondary-hover-background-color;
      box-shadow: $button-secondary-box-shadow;
      color: $button-secondary-hover-color;
      fill: $button-secondary-hover-color;
    }

    &[disabled] {
      background-color: $button-secondary-disabled-background-color;
      box-shadow: $button-secondary-disabled-box-shadow;
      color: $button-secondary-disabled-color;
      cursor: not-allowed;
      fill: $button-secondary-disabled-color;
    }
  }

  /* Size variants
    ========================================================================== */

  /* Large
     ========== */
  &--large {
    @include redwood-display-40;

    padding: $inset-1-and-a-half-x-squish;

    &.cdr-button--has-icon {
      padding-left: calc($space-1-and-a-half-x - $space-quarter-x);
    }
  }

  /* Medium
     ========== */
  &--medium {
    @include redwood-display-30;

    padding: $inset-1-x-squish;

    &.cdr-button--has-icon {
      padding-left: calc($space-1-x - $space-quarter-x);
    }
  }

  /* Small
     ========== */
  &--small {
    @include redwood-display-20;

    padding: $inset-three-quarter-x-squish;

    &.cdr-button--has-icon {
      padding-left: calc($space-three-quarter-x - $space-quarter-x);
    }

    & > :global(.cdr-button__icon) {
      /* This is to correct icon size vs line height */
      @include small-button-icon();
    }
  }

  /* Full Width
     ========== */
  &--full-width {
    width: 100%;
    justify-content: center;
  }

  /* Icon Only
     ========= */

  &--icon-only {
    background: none;
    border: none;
    box-shadow: none;
    display: inline-block;
    padding: $inset-half-x;
    padding-right: calc($inset-half-x - $space-quarter-x);
    fill: $heart-of-darkness;

    &:hover,
    &:active,
    &:focus {
      background: none;
      box-shadow: none;
      fill: $heart-of-darkness;
    }

    &.cdr-button--on-dark {
      fill: $moon-shot;
    }
  }

  /* Breakpoint variants
    ========================================================================== */

  @mixin auto-icon() {
    width: $icon-size;
    height: $icon-size;
  }

  @mixin small-button-icon() {
    width: 20px;
    height: 20px;
  }

  /* @xs
    0px - 767px
    ========== */
  @media (--xs-mq-only) {
    &--small\@xs {
      @include redwood-display-20;

      padding: $inset-three-quarter-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include small-button-icon();
      }
    }

    &--medium\@xs {
      @include redwood-display-30;

      padding: $inset-1-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }

    &--large\@xs {
      @include redwood-display-40;

      padding: $inset-1-and-a-half-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }
  }

  /* @sm
    768px - 991px
    ========== */
  @media (--sm-mq-only) {
    &--small\@sm {
      @include redwood-display-20;

      padding: $inset-three-quarter-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include small-button-icon();
      }
    }

    &--medium\@sm {
      @include redwood-display-30;

      padding: $inset-1-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }

    &--large\@sm {
      @include redwood-display-40;

      padding: $inset-1-and-a-half-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }
  }

  /* @md
    992px - 1199px
    ========== */
  @media (--md-mq-only) {
    &--small\@md {
      @include redwood-display-20;

      padding: $inset-three-quarter-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include small-button-icon();
      }
    }

    &--medium\@md {
      @include redwood-display-30;

      padding: $inset-1-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }

    &--large\@md {
      @include redwood-display-40;

      padding: $inset-1-and-a-half-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }
  }

  /* @lg
    1200px and up
    ========== */

  @media (--lg-mq-only) {
    &--small\@lg {
      @include redwood-display-20;

      padding: $inset-three-quarter-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include small-button-icon();
      }
    }

    &--medium\@lg {
      @include redwood-display-30;

      padding: $inset-1-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }

    &--large\@lg {
      @include redwood-display-40;

      padding: $inset-1-and-a-half-x-squish;
      width: auto;

      & > :global(.cdr-button__icon) {
        @include auto-icon();
      }
    }
  }
}
