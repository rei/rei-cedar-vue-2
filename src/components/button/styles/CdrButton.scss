@import '../../../css/settings/index.scss';
@import './vars/CdrButton.vars.scss';
.cdr-button {
  @include cdr-button-base-mixin;

  & > :global(.cdr-button__icon) {
    @include cdr-button-base-icon-mixin;
  }

  /* Modifier style variants
    ========================================================================== */
  &--primary {
    @include cdr-button-primary-mixin;
  }

  /* Secondary
     ========== */
  &--secondary {
    @include cdr-button-secondary-mixin;
  }

  /* Size variants
    ========================================================================== */

  /* Large
     ========== */
  &--large {
    @include cdr-button-large-mixin;

    &.cdr-button--has-icon {
      @include cdr-button-has-icon-large-mixin;
    }

    & > :global(.cdr-button__icon) {
      @include cdr-button-icon-large-mixin;
    }
  }

  /* Medium
     ========== */
  &--medium {
    @include cdr-button-medium-mixin;

    &.cdr-button--has-icon {
      @include cdr-button-has-icon-medium-mixin;
    }

    & > :global(.cdr-button__icon) {
      @include cdr-button-icon-medium-mixin;
    }
  }

  /* Small
     ========== */

  &--small {
    @include cdr-button-small-mixin;

    &.cdr-button--has-icon {
      @include cdr-button-has-icon-small-mixin;
    }

    & > :global(.cdr-button__icon) {
      @include cdr-button-icon-small-mixin;
    }
  }

  /* Icon Only
     ========= */

  &--icon-only {
    @include cdr-button-icon-only-mixin;

    & > :global(.cdr-button__icon) {
      margin: 0;
    }
  }

  /* Full Width
    ========== */

  &--full-width {
    @include cdr-button-full-width-mixin;
  }

  /* Breakpoint variants

  /* @xs
    0px - 767px
    ========== */
  @include xs-mq-only {
    &--small\@xs {
      @include cdr-button-small-mixin;

      &.cdr-button--has-icon {
        @include cdr-button-has-icon-small-mixin;
      }

      & > :global(.cdr-button__icon) {
        @include cdr-button-icon-small-mixin;
      }
    }

    &--medium\@xs {
      @include cdr-button-medium-mixin;

      &.cdr-button--has-icon {
        @include cdr-button-has-icon-medium-mixin;
      }

      & > :global(.cdr-button__icon) {
        @include cdr-button-icon-medium-mixin;
      }
    }

    &--large\@xs {
      @include cdr-button-large-mixin;

      &.cdr-button--has-icon {
        @include cdr-button-has-icon-large-mixin;
      }

      & > :global(.cdr-button__icon) {
        @include cdr-button-icon-large-mixin;
      }
    }

    &--fullWidth\@xs {
      @include cdr-button-full-width-mixin;
    }
  }

  /* @sm
    768px - 991px
    ========== */
  @include sm-mq-only {
    &--small\@sm {
      @include cdr-button-small-mixin;

      &.cdr-button--has-icon {
        @include cdr-button-has-icon-small-mixin;
      }

      & > :global(.cdr-button__icon) {
        @include cdr-button-icon-small-mixin;
      }
    }

    &--medium\@sm {
      @include cdr-button-medium-mixin;

      &.cdr-button--has-icon {
        @include cdr-button-has-icon-medium-mixin;
      }

      & > :global(.cdr-button__icon) {
        @include cdr-button-icon-medium-mixin;
      }
    }

    &--large\@sm {
      @include cdr-button-large-mixin;

      &.cdr-button--has-icon {
        @include cdr-button-has-icon-large-mixin;
      }

      & > :global(.cdr-button__icon) {
        @include cdr-button-icon-large-mixin;
      }
    }

    &--fullWidth\@sm {
      @include cdr-button-full-width-mixin;
    }
  }

  /* @md
    992px - 1199px
    ========== */
  @include md-mq-only {
    &--small\@md {
      @include cdr-button-small-mixin;

      &.cdr-button--has-icon {
        @include cdr-button-has-icon-small-mixin;
      }

      & > :global(.cdr-button__icon) {
        @include cdr-button-icon-small-mixin;
      }
    }

    &--medium\@md {
      @include cdr-button-medium-mixin;

      &.cdr-button--has-icon {
        @include cdr-button-has-icon-medium-mixin;
      }

      & > :global(.cdr-button__icon) {
        @include cdr-button-icon-medium-mixin;
      }
    }

    &--large\@md {
      @include cdr-button-large-mixin;

      &.cdr-button--has-icon {
        @include cdr-button-has-icon-large-mixin;
      }

      & > :global(.cdr-button__icon) {
        @include cdr-button-icon-large-mixin;
      }
    }

    &--fullWidth\@md {
      @include cdr-button-full-width-mixin;
    }
  }

  /* @lg
    1200px and up
    ========== */

  @include lg-mq-only {
    &--small\@lg {
      @include cdr-button-small-mixin;

      &.cdr-button--has-icon {
        @include cdr-button-has-icon-small-mixin;
      }

      & > :global(.cdr-button__icon) {
        @include cdr-button-icon-small-mixin;
      }
    }

    &--medium\@lg {
      @include cdr-button-medium-mixin;

      &.cdr-button--has-icon {
        @include cdr-button-has-icon-medium-mixin;
      }

      & > :global(.cdr-button__icon) {
        @include cdr-button-icon-medium-mixin;
      }
    }

    &--large\@lg {
      @include cdr-button-large-mixin;

      &.cdr-button--has-icon {
        @include cdr-button-has-icon-large-mixin;
      }

      & > :global(.cdr-button__icon) {
        @include cdr-button-icon-large-mixin;
      }
    }

    &--fullWidth\@lg {
      @include cdr-button-full-width-mixin;
    }
  }
}
