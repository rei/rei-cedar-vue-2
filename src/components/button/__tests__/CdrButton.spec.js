import { shallowMount } from '@vue/test-utils';
import CdrButton from 'componentsdir/button/CdrButton';
import sinon from 'sinon'

describe('CdrButton.vue', () => {
  it('renders a button', () => {
    const wrapper = shallowMount(CdrButton);
    expect(wrapper.is('button')).toBe(true);
  });

  it('sets default type prop correctly', () => {
    const wrapper = shallowMount(CdrButton);
    expect(wrapper.attributes().type).toBe('button');
  });

  it('validates type prop', () => {
    const wrapper = shallowMount(CdrButton);
    const type = wrapper.vm.$options.props.type;
    expect(type.validator('test')).toBe(false);
    expect(type.validator('reset')).toBe(true);
  });

  it('sets type attr correctly', () => {
    const wrapper = shallowMount(CdrButton, {
      propsData: {
        type: 'reset',
      },
    });
    expect(wrapper.attributes().type).toBe('reset');
  });

  it('has default click', () => {
    const wrapper = shallowMount(CdrButton);
    const defaultFunc = wrapper.vm.$props.onClick();
    const result = defaultFunc();
    expect(result).toBe(null)
  });

  it('click function triggers correctly', () => {
    const spy = sinon.spy();
    const wrapper = shallowMount(CdrButton, {
      propsData: {
        onClick: spy
      },
    });
    wrapper.trigger('click');
    expect(spy.calledOnce).toBeTruthy();
  });

  it('computes base class correctly', () => {
    const wrapper = shallowMount(CdrButton, {
      propsData: {
        el: 'a',
      },
    });
    
    expect(wrapper.classes()).toContain('cdr-button');
  });

  it('adds responsive classes', () => {
    const wrapper = shallowMount(CdrButton, {
      propsData: {
        staticSize: 'medium',
        responsiveSize: ['large@xs'],
      }
    });

    expect(wrapper.vm.$style).toHaveProperty(`cdr-button--large@xs`);
  })

  it('does not add size class when icon only', () => {
    const wrapper = shallowMount(CdrButton, {
      propsData: {
        size: 'large',
        iconOnly: true,
      },
    });
    expect(wrapper.vm.sizeClass).toBe(null);
  })

  it('does not add icon class when slot is unused', () => {
    const wrapper = shallowMount(CdrButton);
    expect(wrapper.vm.iconClass).toBe('');
  });

  it('adds classes for icon only and on-dark', () => {
    const wrapper = shallowMount(CdrButton, {
      propsData: {
        iconOnly: true,
        onDark: true,
      }
    });
    expect(wrapper.classes()).toContain('cdr-button--on-dark');
    expect(wrapper.classes()).toContain('cdr-button--icon-only');
  });

  it('adds icon class when slot is used', () => {
    const wrapper = shallowMount(CdrButton, {
      slots: {
        default: [
          'named slot icon',
          'default slot for text'
        ],
      },
    });

    expect(wrapper.classes()).toContain('cdr-button--has-icon');
  });

  it('renders a link', () => {
    const wrapper = shallowMount(CdrButton, {
      propsData: {
        tag: 'a',
      }, 
    });
    expect(wrapper.is('a')).toBe(true);
  });

  it('validates el prop', () => {
    const wrapper = shallowMount(CdrButton);
    const tag = wrapper.vm.$options.props.tag;
    expect(tag.validator('button')).toBe(true);
    expect(tag.validator('link')).toBe(false);
  });

  it('validates size prop', () => {
    const wrapper = shallowMount(CdrButton);
    const size = wrapper.vm.$options.props.size;
    expect(size.validator('small')).toBe(true);
    expect(size.validator('extra-small')).toBe(false);
  });
});