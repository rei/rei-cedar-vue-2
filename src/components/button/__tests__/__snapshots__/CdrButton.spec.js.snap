// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrButton renders correctly 1`] = `
<button
  class="cdr-button--primary cdr-button cdr-button cdr-button--medium"
  type="button"
/>
`;
